### PR TITLE
EU-90

### PR DESCRIPTION
**Pull request by victor-nuzhniy**
Waiting for callback. Pleas look at the get_bed_places and post_university_id_faculty endpoints, added to them info using function docstring. Of course info repeat already present in the page, but with little more information. If it looks ok, can add such description to all endpoints.

### CODE REVIEW

**Review by anton-serozhechkin — CHANGES_REQUESTED** [review]

File: handlers/user_request.py [hunk 1]
```diff
@@ -127,12 +150,16 @@ async def cancel_request(university_id: int, user_request_id: int, cancel_reques
...
+        "message": f"Canceled request with id {user_request_id}",
+        "code": http_status.HTTP_202_ACCEPTED
     }
 
 
 @router.post("/{university_id}/user-request-review/{user_request_id}/",
+             name="post_user_request_review",
              response_model=JSENDOutSchema[UserRequestReviewOut],
+             summary="Create user request review",
+             responses={200: {"description": "Successful create user request review response"}},
```
Comment (anton-serozhechkin) [inline handlers/user_request.py:162]: Successfully created
Please delete response, i thnk it's useless

File: handlers/authorization/registration.py [hunk 2]
```diff
@@ -14,13 +14,23 @@
 from translitua import translit
 from fastapi import APIRouter, status as http_status
 
-from schemas.jsend import JSENDOutSchema
+from schemas.jsend import JSENDOutSchema, JSENDErrorOutSchema, JSENDFailOutSchema
 from components.exceptions import BackendException
 
-router = APIRouter()
+router = APIRouter(
+    responses={422: {"model": JSENDErrorOutSchema, "description": "ValidationError"},
```
Comment (Kostiantyn-Salnykov) [inline handlers/authorization/registration.py:21]: Change to Fail

File: handlers/hostel.py [hunk 3]
```diff
@@ -9,13 +9,20 @@
...
 from fastapi import APIRouter, Depends
 
-from schemas.jsend import JSENDOutSchema
+from schemas.jsend import JSENDOutSchema, JSENDErrorOutSchema
 
 
-router = APIRouter()
+router = APIRouter(
+    tags=["Admin dashboard"],
+    responses={422: {"model": JSENDErrorOutSchema, "description": "ValidationError"}}
```
Comment (Kostiantyn-Salnykov) [inline handlers/hostel.py:17]: Change to Fail

File: handlers/speciality.py [hunk 4]
```diff
@@ -9,14 +9,20 @@
...
 from fastapi import APIRouter, Depends
 
-from schemas.jsend import JSENDOutSchema
+from schemas.jsend import JSENDOutSchema, JSENDErrorOutSchema
 
 
-router = APIRouter()
+router = APIRouter(
+    tags=["Admin dashboard"],
+    responses={422: {"model": JSENDErrorOutSchema, "description": "ValidationError"}}
```
Comment (Kostiantyn-Salnykov) [inline handlers/speciality.py:17]: Change to Fail

File: handlers/student.py [hunk 5]
```diff
@@ -8,14 +8,21 @@
...
 
-from fastapi import Depends, APIRouter
+from fastapi import APIRouter, Depends, status as http_status
 
-from schemas.jsend import JSENDOutSchema
+from schemas.jsend import JSENDOutSchema, JSENDErrorOutSchema
 
-router = APIRouter()
+router = APIRouter(
+    responses={422: {"model": JSENDErrorOutSchema, "description": "ValidationError"}}
```
Comment (Kostiantyn-Salnykov) [inline handlers/student.py:16]: Change to Fail

File: handlers/user_request.py [hunk 6]
```diff
@@ -22,15 +22,21 @@
...
 
-from fastapi import Depends, APIRouter
+from fastapi import Depends, APIRouter, status as http_status
 
-from schemas.jsend import JSENDOutSchema
+from schemas.jsend import JSENDOutSchema, JSENDErrorOutSchema
 
-router = APIRouter()
+router = APIRouter(
+    responses={422: {"model": JSENDErrorOutSchema, "description": "ValidationError"}}
```
Comment (Kostiantyn-Salnykov) [inline handlers/user_request.py:30]: Change to Fail

File: schemas/jsend.py [hunk 7]
```diff
@@ -15,3 +15,15 @@ class JSENDOutSchema(GenericModel, Generic[SchemaVar]):
...
+class JSENDFailOutSchema(JSENDOutSchema):
+    status: JSENDStatus = Field(default=JSENDStatus.FAIL)
+    data: Union[str, NoneType]
+    code: int = Field(default=http_status.HTTP_400_BAD_REQUEST)
+
+
+class JSENDErrorOutSchema(JSENDOutSchema):
+    status: JSENDStatus = Field(default=JSENDStatus.ERROR)
+    data: Union[str, NoneType]
+    code: int = Field(default=http_status.HTTP_400_BAD_REQUEST)
```
Comment (Kostiantyn-Salnykov) [inline schemas/jsend.py:29]: Should be a `500` error code or custom (in the future)

**Review by Kostiantyn-Salnykov — CHANGES_REQUESTED** [review]

File: handlers/authorization/auth.py [hunk 8]
```diff
@@ -4,28 +4,39 @@
 from schemas.user import AuthOut
 
 from sqlalchemy import select
-from fastapi import Depends, APIRouter
+from fastapi import Depends, APIRouter, status as http_status
```
[the remark this answers is not in the document]
Comment (victor-nuzhniy) [reply handlers/authorization/auth.py:7]: Added http_... in previous pr, may be after somebodys advise. Is it wrong? It seems to me the code readability is better

[the remark this answers is not in the document]
Comment (Kostiantyn-Salnykov) [reply handlers/authorization/auth.py:7]: OK, left it as it is

File: apps/services/handlers.py [hunk 9]
```diff
@@ -54,9 +77,25 @@ async def read_user_request_list(university_id: int, user=Depends(get_current_us
...
 async def create_user_request(university_id: int, user_request: CreateUserRequestIn, user=Depends(get_current_user)):
+    """
+    **Create user request**
+
+    **Path**:
+    - **university_id**: user university id
+
+    **Input**:
+    - **service_id**: service id in database, required
+    - **comment*: comment for the creating user request, required
```
Comment (anton-serozhechkin) [inline apps/services/handlers.py]: comment isn't required

File: apps/services/handlers.py [hunk 10]
```diff
@@ -102,10 +146,26 @@ async def read_user_request_booking_hostel(university_id: int, user=Depends(get_
...
+    **Cancel user request**
+
+    **Path**:
+    - **university_id**: user university id
+    - **user_request_id**: user request id
+
+    **Input**:
+    - **status_id**: user request status id, required
+
+    **Return**: deleted user request id and status id
```
Comment (anton-serozhechkin) [inline apps/services/handlers.py]: not deleted, cancelled

File: apps/services/handlers.py [hunk 11]
```diff
@@ -102,10 +146,26 @@ async def read_user_request_booking_hostel(university_id: int, user=Depends(get_
     }
 
 
-@services_router.put("/{university_id}/user-request/{user_request_id}", response_model=JSENDOutSchema[CancelRequestOut],
+@services_router.put("/{university_id}/user-request/{user_request_id}",
+                     name="update_cancel_user_request",
```
Comment (anton-serozhechkin) [inline apps/services/handlers.py:150]: why update?

File: apps/services/handlers.py [hunk 12]
```diff
@@ -116,15 +176,39 @@ async def cancel_request(university_id: int, user_request_id: int, cancel_reques
             "user_request_id": user_request_id,
             "status_id": cancel_request.status_id
         },
-        "message": f"Canceled request with id {user_request_id}"
+        "message": f"Canceled request with id {user_request_id}",
+        "code": http_status.HTTP_202_ACCEPTED
```
Comment (anton-serozhechkin) [inline apps/services/handlers.py]: im not sure that 202 is correct

File: apps/services/handlers.py [hunk 13]
```diff
@@ -116,15 +176,39 @@ async def cancel_request(university_id: int, user_request_id: int, cancel_reques
...
+    **Input**:
+    - **status_id**: user request status id, required
+    - **room_number**: user room number, required
+    - **start_date_accommodation**: starting datetime hostel accommodation, required
+    - **end_date_accommodation**: end datetime hostel accommodation, required
+    - **total_sum**: total sum of hostel accommodation payment, required
+    - **payment_deadline**: deadline datetime for hostel accommodation payment, required
+    - **remark**: additional info for request review, required
+    - **hostel_id**: hostel id in the database, required
+    - **bed_place_id**: hostel bed place id, required
```
Comment (anton-serozhechkin) [inline apps/services/handlers.py:208]: more of these fields aren't required

File: apps/services/handlers.py (hunk 11 above)
Comment (victor-nuzhniy) [reply apps/services/handlers.py:150]: CRUD, that's why. Of course, I can change, but only one (or two) will differ from others. What to do?

File: apps/users/handlers.py [hunk 14]
```diff
@@ -66,10 +68,26 @@ async def get_current_user(token: str = Depends(reuseable_oauth)) -> UserOut:
     return user
 
 
-@users_router.post("/check-student-existance", response_model=JSENDOutSchema[StudentCheckExistanceOut],
+@users_router.post("/check-student-existance",
+                   name="create_student_existence",
```
Comment (anton-serozhechkin) [inline apps/users/handlers.py:72]: why create?

File: apps/services/handlers.py (hunk 12 above)
Comment (victor-nuzhniy) [reply apps/services/handlers.py]: Couldn't find 100 % appropriate, but google says its ok. ?

File: apps/users/handlers.py [hunk 15]
```diff
@@ -112,8 +137,31 @@ async def users_list(university_id: int, user=Depends(get_current_user)):
...
+        **Create university user**
+
+        **Path**:
+        - **university_id**: university id for creating user
+
+        **Input**:
+        - **email**: user email, only letters (a-z), numbers (0-9) and periods (.) are allowed, required
+        - **password**: password, cannot be empty, required
+        - **password_re_check**: password recheck, required
+        - **role_id**: user role id, 1 - student, 2 - admin, 3 - superadmin, required
```
Comment (anton-serozhechkin) [inline apps/users/handlers.py]: it's not a good idea to describe role values there is bcs it can be changed lately

File: apps/users/handlers.py [hunk 16]
```diff
@@ -155,13 +212,34 @@ async def delete_user(university_id: int, delete_user: DeleteUserIn, auth=Depend
         "data": {
             "user_id": delete_user.user_id
         },
-        "message": f"Deleted user with id {delete_user.user_id}"
+        "message": f"Deleted user with id {delete_user.user_id}",
+        "code": http_status.HTTP_202_ACCEPTED
```
Comment (anton-serozhechkin) [inline apps/users/handlers.py]: For a DELETE request: HTTP 200 or HTTP 204 should imply "resource deleted successfully".

HTTP 202 can also be returned by either operation and would imply that the instruction was accepted by the server, but not fully applied yet.

File: apps/services/handlers.py (hunk 13 above)
Comment (victor-nuzhniy) [reply apps/services/handlers.py:208]: I've asked about it Konstantin, he said that we can add such info in cases with a great amount different data, that need to be described. Of course I can delete it, but first would be a good idea to see how it looks. Waiting for further info what to do.

File: apps/users/handlers.py (hunk 14 above)
Comment (victor-nuzhniy) [reply apps/users/handlers.py:72]: CRUD

File: apps/users/handlers.py (hunk 15 above)
Comment (victor-nuzhniy) [reply apps/users/handlers.py]: We can everything change lately

File: apps/users/handlers.py (hunk 16 above)
Comment (victor-nuzhniy) [reply apps/users/handlers.py]: Ok. What code should I use?

File: apps/users/handlers.py [hunk 17]
```diff
@@ -276,24 +389,37 @@ async def read_students_list(university_id: int, faculty_id: Union[int, None] =
...
     await database.execute(query)
-    # TODO: in response key data has empty dict value, not like it's discribed
+    # TODO: in response key data has empty dict value, not like it's described
     return {
         "data": {
             "student_id": delete_student.student_id
         },
-        "message": f"Deleted student with id {delete_student.student_id}"
+        "message": f"Deleted student with id {delete_student.student_id}",
+        "code": http_status.HTTP_202_ACCEPTED
```
Comment (anton-serozhechkin) [inline apps/users/handlers.py]: For a DELETE request: HTTP 200 or HTTP 204 should imply "resource deleted successfully".

HTTP 202 can also be returned by either operation and would imply that the instruction was accepted by the server, but not fully applied yet.

Comment (victor-nuzhniy) [reply apps/users/handlers.py]: 204?

Comment (anton-serozhechkin) [reply apps/users/handlers.py]: yes

File: apps/users/handlers.py (hunk 16 above)
Comment (anton-serozhechkin) [reply apps/users/handlers.py]: 204

File: apps/users/handlers.py (hunk 15 above)
Comment (anton-serozhechkin) [reply apps/users/handlers.py]: sure, but anyway, data must store in db, not in api
please, delete list of roles

File: apps/services/handlers.py (hunk 13 above)
Comment (anton-serozhechkin) [reply apps/services/handlers.py:208]: ok ,ok for now

File: apps/services/handlers.py (hunk 11 above)
Comment (anton-serozhechkin) [reply apps/services/handlers.py:150]: ok, for now it's ok

File: apps/services/handlers.py (hunk 12 above)
Comment (anton-serozhechkin) [reply apps/services/handlers.py]: For a PUT request: HTTP 200, HTTP 204 should imply "resource updated successfully". HTTP 201 if the PUT request created a new resource.

so - 200 is ok for this example

File: apps/users/handlers.py (hunk 14 above)
Comment (anton-serozhechkin) [reply apps/users/handlers.py:72]: ok

File: apps/users/handlers.py (hunk 16 above)
Comment (anton-serozhechkin) [reply apps/users/handlers.py]: sorry, change it to 200

File: apps/users/handlers.py (hunk 17 above)
Comment (anton-serozhechkin) [reply apps/users/handlers.py]: sorry change it to 2000

**Review by anton-serozhechkin — APPROVED** [review]

**Pull request merged**